README.md
# REST-API-component

This is the Rest api component adjusted to the new ferryman library

The **REST API component** is a simple yet powerful component that allows you to connect to any REST API without programming your own components and deploying them into the platform.

The REST API component will perform a single REST API call when executed. Incoming data can gets used to configure the API call made and the response from the API call will be the output.

This document covers the following topics:

- [Introduction](#introduction)
- [Authorization methods](#authorisation-methods)
- [Defining HTTP headers](#defining-http-headers)
- [HTTP Headers in Response](#http-headers)

## Introduction

_Numbers show: (1) The URL and method of the REST API resource, (2) the HTTP call headers. (3) configuration options and (4) follow redirect mode._

1.  HTTP methods and URL

- REST API component supports the following HTTP methods: `GET`, `PUT`, `POST`, `DELETE` and `PATCH`.
- The URL of the REST API resources. Accepts JSONata expressions, meaning the URL address evaluates [JSONata](http://jsonata.org/) expressions.

2. Request Headers and Body

- Definition of request [headers](#defining-http-headers)
- Definition of request [body](#defining-http-body), if the HTTP method is not `GET`

3. Configuration options

- `` Don`t throw Error on Failed Calls `` - if enabled return error, error code and stacktrace in message body otherwise throw error in flow.
- `Split Result if it is an Array` - if enabled and response is array, creates message for each item of array. Otherwise create one message with response array.
- `Retry on failure` - enabling [rebound] feature for following HTTP status codes:
  - 408: Request Timeout
  - 423: Locked
  - 429: Too Many Requests
  - 500: Internal Server Error
  - 502: Bad Gateway
  - 503: Service Unavailable
  - 504: Gateway Timeout
  - DNS lookup timeout

4. `Do not verify SSL certificate (unsafe)` - disable verifying the server certificate - **unsafe**.
5. `Follow redirect mode` - If you want disable Follow Redirect functionality, you can use option `Follow redirect mode`.By default `Follow redirect mode` option has value `Follow redirects`.
6. `Delay` - If you want to slow down requests to your API you can set delay value (in seconds) and the component will delay calling the next request after the previous request.
   Time for the delay is calculated as `Delay`/ `Call Count` and shouldn't be more than 1140 seconds (19 minutes due to platform limitation).
   The `Call Count` value by default is 1. If you want to use another value, please set the `Call Count` field.
   Notice: See [Known Limitations](#known-limitations) about `Delay` value.
7. `Call Count` - the field should be used only in pair with `Delay`, default to 1.
8. `Request timeout` - Timeout period in milliseconds (1-1140000) while component waiting for server response, also can be configured with REQUEST_TIMEOUT environment variable if configuration field is not provided. Defaults to 100000 (100 sec).
   Notice: Specified for component REQUEST_TIMEOUT enviroment variable would be overwritten by specified value of Request timeout, default value would be also overwritten

## Authorisation methods

To use the REST API component with any restricted access API provide the authorisation information.

"REST API component Basic authorisation"
_Example above shows how to add the username/password to access the API during the integration flow design._

You can add the authorisation methods during the integration flow design or by going to your `Settings > Security credentials > REST client` and adding there.

REST API component supports 4 authorisation types:

- `No Auth` - use this method to work with any open REST API
- `Basic Auth` - use it to provide login credentials like **username/password**
- `API Key Auth` - use it to provide `API Key` to access the resource
- `OAuth2` - use it to provide `Oauth2` credentials to access the resource. Currently it is implemented `Authorization code` OAuth2 flow.

Please note that the result of creating a credential is an HTTP header automatically placed for you. You can also specify the authorisation in the headers section directly.

### Sending JSON data

Here is how to send a JSON data in the body. Change the **content type** to `application/json` and the **body input part** would change accordingly to accept JSON object. Please note that this field supports [JSONata](http://jsonata.org) expressions.

"REST API component Body sending JSON data"
_Example shows the JSON in the body where the `name` parameter value gets mapped using the value of `project_name` from the previous step of integration._

## HTTP Headers

You can to get HTTP response header only if `` Don`t throw Error on Failed Calls `` option is checked.
In this case output structure of component will be:

```js
    {
      headers:<HTTP headers>,
      body:<HTTP response body>,
      statusCode:<HTTP response status code>
      statusMessage:<HTTP response status message>
    }
```


CHANGELOG.md
## 1.2.12 (September 25, 2020)
* Revert changes to url encoding

## 1.2.11 (September 6, 2020)
* Fix url encoding

## 1.2.10 (August 25, 2020)
* Fix `application/x-www-form-urlencoded` encoding bug

## 1.2.9 (July 10, 2020)
* Timeout configuration field

## 1.2.8 (July 6, 2020)
* Add configuration fields: `Delay` and `Call Count` for rate limit
* Update sailor version to 2.6.13

## 1.2.7 (June 24, 2020)
* Add checkBox to not verify servers certificate

## 1.2.6 (June 6, 2020)
* Remove update docs on deploy script

## 1.2.5 (May 19, 2020)
* Update sailor version to 2.6.7

## 1.2.4 (April 9, 2020)
* Fix `No refresh tokens were returned by the OAuth2 provider` error for credentials with non-expiring refresh_token

## 1.2.3 (April 1, 2020)
* New Jsonata expressions support: getFlowVariables and getPathrough

## 1.2.2 (March 19, 2020)
* Add binary message types to list for processing them as attachments

## 1.2.1(March 12, 2020)

### Added
* Add validity check of `refresh_token` in keys for `OAuth2` authentication strategy

## 1.2.0(February 13, 2020)

### Added
* Added new authentication strategy `OAuth2`

## 1.1.4 (December 23, 2019)

* Update sailor version to 2.5.4
* Remove Enable debugging checkbox

## 1.1.3 (December 07, 2019)

* Update sailor version to 2.5.1
* Update jsonata-moment to 1.1.4

## 1.1.2 (September 06, 2019)

* Fix `new Buffer()` deprecated warning

## 1.1.0 (August 28, 2019)

### Added
* Added feature `Retry on failure`

## 1.0.1 (June 26, 2019)

* Initial release which includes a bunch of previous unversioned releases


lib/utils.js
/* eslint-disable max-len,no-shadow,no-param-reassign,no-underscore-dangle,no-use-before-define,consistent-return,arrow-parens */

const { transform } = require("@openintegrationhub/ferryman");

const { newMessage } = require("./messages");

const axios = require("axios");

const {
  getRateLimitDelay,
  rateLimit,
  getRequestTimeout,
  encodeWWWFormParam,
} = require("./helpers");

const HTTP_ERROR_CODE_REBOUND = new Set([408, 423, 429, 500, 502, 503, 504]);

const methodsMap = {
  DELETE: "delete",
  GET: "get",
  PATCH: "patch",
  POST: "post",
  PUT: "put",
};

const bodyEncodings = {
  FORM_DATA: "form-data",
  RAW: "raw",
  URLENCODED: "urlencoded",
};

const bodyMultipartBoundary = "__X_ELASTICIO_BOUNDARY__";

const contentTypes = {
  FORM_DATA: "multipart/form-data",
  URLENCODED: "application/x-www-form-urlencoded",
  TEXT: "text/plain",
  APP_JSON: "application/json",
  APP_XML: "application/xml",
  TEXT_XML: "text/xml",
  HTML: "text/html",
};

const formattedFormDataHeader = `multipart/form-data; charset=utf8; boundary=${bodyMultipartBoundary}`;

const authTypes = {
  NO_AUTH: "No Auth",
  BASIC: "Basic Auth",
  API_KEY: "API Key Auth",
  OAUTH2: "OAuth2",
};

const CREDS_HEADER_TYPE = "CREDS_HEADER_TYPE";

/**
 * Executes the action's/trigger's logic by sending a request to the assigned URL and emitting response to the platform.
 * The function returns a Promise sending a request and resolving the response as platform message.
 *
 * @param {Object} msg incoming messages which is empty for triggers
 * @param {Object} cfg object to retrieve triggers configuration values, such as, for example, url and userId
 * @returns {Object} promise resolving a message to be emitted to the platform
 */
/* eslint-disable-next-line func-names */
module.exports.processMethod = async function (msg, cfg) {
  const emitter = this;

  emitter.logger.debug("Input message: %o", JSON.stringify(msg));
  emitter.logger.debug("Input configuration: %o", JSON.stringify(cfg));

  const config = cfg.reader;

  if (!config.url) {
    console("there is no url", config.url);
    throw new Error("URL is required");
  }
  console.log("this is the transformed url:", config.url);

  const { method, headers, token } = config;
  const body = config.body || {};
  const followRedirect = cfg.followRedirect !== "doNotFollowRedirects";
  const { auth } = cfg;
  const requestTimeout = getRequestTimeout(emitter.logger, cfg);
  // const token = cfg.token;

  if (!method) {
    throw new Error("Method is required");
  }

  const formattedMethod = methodsMap[method];

  if (!formattedMethod) {
    throw new Error(
      `Method "${method}" isn't one of the: ${Object.keys(methodsMap)}.`
    );
  }

  const rateLimitDelay = getRateLimitDelay(emitter.logger, cfg);

  /*
   if cfg.followRedirect has value doNotFollowRedirects
   or cfg.followRedirect is not exists
   followRedirect option should be true
   */
  const requestOptions = {
    method: formattedMethod,
    url: config.url,
    followRedirect,
    followAllRedirects: followRedirect,
    gzip: true,
    resolveWithFullResponse: true,
    simple: false,
    encoding: null,
    strictSSL: !cfg.noStrictSSL,
    timeout: requestTimeout,
    headers: {
      Authorization: "Bearer " + token,
    },
    body: body,
    data: body,
  };

  console.log("these are the requestOptions:", requestOptions);

  switch (auth.type) {
    case authTypes.BASIC:
      headers.push({
        key: "Authorization",
        // eslint-disable-next-line no-buffer-constructor
        value: `"Basic ${Buffer.from(
          `${auth.basic.username}:${auth.basic.password}`,
          "utf8"
        ).toString("base64")}"`,
      });

      break;

    case authTypes.API_KEY:
      headers.push({
        key: auth.apiKey.headerName,
        value: `"${auth.apiKey.headerValue}"`,
      });

      break;
    case authTypes.OAUTH2:
      emitter.logger.trace("auth = %j", auth);
      // eslint-disable-next-line no-case-declarations

      headers.push({
        key: "Authorization",
        value: `"Bearer ${token}"`,
      });
      break;
    default:
  }

  if (headers && headers.length) {
    requestOptions.headers = headers.reduce((headers, header) => {
      if (!header.key || !header.value) {
        return headers;
      }
      headers[header.key.toLowerCase()] = transform(
        msg,
        { expression: header.value },
        emitter
      );
      return headers;
    }, requestOptions.headers || {});
  }

  emitter.logger.debug("Request options: %o", JSON.stringify(requestOptions));

  return buildRequestBody()
    .then(() => {
      emitter.logger.trace("Request body: %o", requestOptions.body);
      console.log(requestOptions.body);
      return axios(requestOptions);
    })
    .then(async (result) => {
      console.log(result);
      console.log(result.data);
      emitter.logger.trace("Request output: %j", result);

      if (cfg.splitResult && Array.isArray(result)) {
        // Walk through chain of promises: https://stackoverflow.com/questions/30445543/execute-native-js-promise-in-series
        // eslint-disable-next-line no-restricted-syntax
        for (const item of result) {
          const output = newMessage(item);
          // eslint-disable-next-line no-await-in-loop
          await emitter.emit("data", output);
        }
      } else {
        const output = newMessage(result);
        await emitter.emit("data", output);
      }
      await rateLimit(emitter.logger, rateLimitDelay);
      await emitter.emit("end");
    })
    .catch(await buildErrorStructure);

  function checkOAuth2Keys(keys) {
    emitter.logger.trace("Check keys = %j", keys);
    if (!keys) {
      throw new Error("cfg.auth.oauth2.keys can not be empty");
    }
    if (!keys.access_token) {
      throw new Error("No access tokens were returned by the OAuth2 provider");
    }
    if (!keys.refresh_token) {
      throw new Error(
        "No refresh tokens were returned by the OAuth2 provider. Try to add access_type:offline as an additional parameter"
      );
    }
  }

  async function fetchNewToken() {
    emitter.logger.debug("Fetching new oauth2 token...");
    const { oauth2 } = auth;
    const authTokenResponse = await axios({
      url: oauth2.tokenUri,
      method: "POST",
      json: true,
      simple: false,
      resolveWithFullResponse: true,
      form: {
        refresh_token: oauth2.keys.refresh_token,
        grant_type: "refresh_token",
        client_id: oauth2.clientId,
        client_secret: oauth2.clientSecret,
        scope: oauth2.scopes ? oauth2.scopes.join(" ") : "",
      },
    });

    emitter.logger.trace("New token fetched : %j", authTokenResponse);

    if (authTokenResponse.statusCode >= 400) {
      throw new Error(
        `Error in authentication.  Status code: ${
          authTokenResponse.statusCode
        }, Body: ${JSON.stringify(authTokenResponse.body)}`
      );
    }

    return authTokenResponse.body;
  }

  function buildRequestBody() {
    if (formattedMethod !== methodsMap.GET) {
      const bodyEncoding =
        {
          [contentTypes.FORM_DATA]: bodyEncodings.FORM_DATA,
          [contentTypes.URLENCODED]: bodyEncodings.URLENCODED,
        }[body.contentType] || bodyEncodings.RAW;

      // eslint-disable-next-line default-case
      switch (bodyEncoding) {
        case bodyEncodings.FORM_DATA:
          // eslint-disable-next-line no-case-declarations
          const existingContentTypeHeader = headers.find(
            (header) => (
              // eslint-disable-next-line no-sequences
              header.key.match(/^content-type$/i),
              header.value === contentTypes.FORM_DATA
            )
          );

          if (existingContentTypeHeader) {
            existingContentTypeHeader.value = `"${formattedFormDataHeader}"`;
          } else {
            headers.push({
              key: "Content-Type",
              value: `"${formattedFormDataHeader}"`,
            });
          }

          emitter.logger.trace("formData: %o", body.formData);

          requestOptions.body = `--${bodyMultipartBoundary}`;

          return body.formData
            .reduce((p, x) => p.then(() => processItem(x)), Promise.resolve())
            .then(() => {
              requestOptions.body = `${requestOptions.body}--`;
              return requestOptions.body;
            });

        case bodyEncodings.RAW:
          if (!body.raw) {
            break;
          }

          requestOptions.body = transform(
            msg,
            { expression: body.raw },
            emitter
          );
          if (typeof requestOptions.body === "object") {
            requestOptions.body = JSON.stringify(requestOptions.body);
          }
          break;

        case bodyEncodings.URLENCODED:
          if (!body.urlencoded.length) {
            break;
          }

          // eslint-disable-next-line no-case-declarations
          const evaluatedUrlencoded = body.urlencoded
            .map((pair) => ({
              key: pair.key,
              value: transform(msg, { expression: pair.value }, emitter),
            }))
            .reduce((str, pair, index) => {
              const equation = `${encodeWWWFormParam(
                pair.key
              )}=${encodeWWWFormParam(pair.value)}`;

              return index === 0 ? equation : `${str}&${equation}`;
            }, null);

          requestOptions.body = evaluatedUrlencoded;
          break;
      }
      emitter.logger.trace("Request body: %o", requestOptions.body);
    }

    function processItem(item) {
      if (item.filename) {
        return axios(item.value)
          .then((result) => {
            requestOptions.body = `${requestOptions.body}\nContent-Disposition: form-data; name="${item.key}"; filename:"${item.filename}"\nContent-Type:${item["Content-Type"]}\n\n${result}\n--${bodyMultipartBoundary}`;
          })
          .catch((result) => {
            emitter.logger.trace(result);
          });
      }
      return Promise.resolve().then(() => {
        requestOptions.body =
          `${requestOptions.body}\nContent-Disposition: form-data; name="${item.key}"\n\n` +
          `${transform(
            msg,
            { expression: item.value },
            emitter
          )}\n--${bodyMultipartBoundary}`;
      });
    }

    return Promise.resolve(requestOptions.body);
  }

  async function buildErrorStructure(e) {
    if (
      cfg.enableRebound &&
      (HTTP_ERROR_CODE_REBOUND.has(e.code) ||
        e.message.includes("DNS lookup timeout"))
    ) {
      emitter.logger.info("Component error: %o", e);
      emitter.logger.info("Starting rebound");
      emitter.emit("rebound", e.message);
      emitter.emit("end");
    } else if (cfg.dontThrowErrorFlg) {
      const output = {
        errorCode: e.code,
        errorMessage: e.message,
        errorStack: e.stack,
      };
      emitter.logger.debug("Component output: %o", output);
      await emitter.emit("data", newMessage(output));
      await rateLimit(emitter.logger, rateLimitDelay);
      await emitter.emit("end");
    } else {
      emitter.logger.error("Component error: %o", e);
      if (e.message === "Error: ESOCKETTIMEDOUT") {
        e.message = `Timeout error! Waiting for response more than ${requestTimeout} ms`;
      }
      await emitter.emit("error", e);
      await rateLimit(emitter.logger, rateLimitDelay);
      await emitter.emit("end");
    }
  }
};
